internal/core/src/segcore/ChunkedSegmentSealedImpl.cpp

```diff
@@ -1182,72 +1183,28 @@ ChunkedSegmentSealedImpl::check_search(const query::Plan* plan) const {
-    auto pk_column = fields_.at(pk_field_id);
-    std::vector<SegOffset> pk_offsets;
-    switch (schema_->get_fields().at(pk_field_id).get_data_type()) {
-        case DataType::INT64: {
-            auto target = std::get<int64_t>(pk);
-            // get int64 pks
-            auto num_chunk = pk_column->num_chunks();
-            for (int i = 0; i < num_chunk; ++i) {
-                auto src = reinterpret_cast<const int64_t*>(pk_column->Data(i));
-                auto chunk_row_num = pk_column->chunk_row_nums(i);
-                auto it = std::lower_bound(
-                    src,
-                    src + chunk_row_num,
-                    target,
-                    [](const int64_t& elem, const int64_t& value) {
-                        return elem < value;
-                    });
-                for (; it != src + chunk_row_num && *it == target; it++) {
-                    auto offset = it - src;
-                    if (insert_record_.timestamps_[offset] <= timestamp) {
-                        pk_offsets.emplace_back(offset);
-                    }
-                }
-            }
-            break;
-        }
-        case DataType::VARCHAR: {
-            auto target = std::get<std::string>(pk);
-            // get varchar pks
-            auto var_column =
-                std::dynamic_pointer_cast<ChunkedVariableColumn<std::string>>(
-                    pk_column);
-            auto num_chunk = var_column->num_chunks();
-            for (int i = 0; i < num_chunk; ++i) {
-                auto views = var_column->StringViews(i).first;
-                auto it = std::lower_bound(views.begin(), views.end(), target);
-                for (; it != views.end() && *it == target; it++) {
-                    auto offset = std::distance(views.begin(), it);
-                    if (insert_record_.timestamps_[offset] <= timestamp) {
-                        pk_offsets.emplace_back(offset);
-                    }
-                }
-            }
-            break;
-        }
-        default: {
-            PanicInfo(
-                DataTypeInvalid,
-                fmt::format(
-                    "unsupported type {}",
-                    schema_->get_fields().at(pk_field_id).get_data_type()));
-        }
-    }
-
-    return pk_offsets;
+    return search_pk_internal(pk, [this, timestamp](int64_t offset) {
+        return insert_record_.timestamps_[offset] <= timestamp;
+    });
 }
 
 std::vector<SegOffset>
 ChunkedSegmentSealedImpl::search_pk(const PkType& pk,
                                     int64_t insert_barrier) const {
+    return search_pk_internal(pk, [insert_barrier](int64_t offset) {
+        return offset < insert_barrier;
+    });
+}
+
+template <typename Condition>
+std::vector<SegOffset>
+ChunkedSegmentSealedImpl::search_pk_internal(const PkType& pk,
+                                             Condition condition) const {
     auto pk_field_id = schema_->get_primary_field_id().value_or(FieldId(-1));
```

Comment:
can we just check isSort function here and fallback to insert_record instead of handle everything outside?

Reply:
I did this in search_pk. Only sorted segments will enter this section, so I renamed the function.